### ISSUE TITLE
[CRASH] If you have version.script file,  make sure update Redis* to Valkey*

### ISSUE THREAD

**Issue by hwware** [madolson/placeholderkv#547]
Test steps:  

1. Create a new module hellovalkey.c file under $ValkeyPath/module/hellovalkey folder

[hellovalkey.zip](https://github.com/valkey-io/valkey/files/15472055/hellovalkey.zip)

2. Create a version.script file
 
module {
  global: RedisModule_OnLoad;
  local: *;
};

3. Create a Makefile

CC=cc
CFLAGS=-g -O2 -fPIC  -Wall -shared
LDFLAGS=
LIBS=

all::
        ${CC} ${CFLAGS} -I../../src -o hellovalkey.so hellovalkey.c ${LDFLAGS} ${LIBS} -Wl,--version-script=version.script

clean::
        rm -f *.so *.o

4. compile the module, and we can get hellovalkey.so
5. When you load the hellovalkey module, you will get the following error in server side:

433050:M 28 May 2024 15:18:15.582 # Module /home/ubuntu/valkey-oss/module/hellovalkey/hellovalkey.so does not export ValkeyModule_OnLoad() or RedisModule_OnLoad() symbol. Module not loaded.

and you will get the following error in client side:

172.25.0.58:7000> module load /home/ubuntu/valkey-oss/module/hellovalkey/hellovalkey.so
(error) ERR Error loading the extension. Please check the server logs.

6.  Update the version.script as following:

module {
  global: ValkeyModule_OnLoad;
  local: *;
};

7. compile the hellovalkey module again, and you will get the correct result

![image](https://github.com/valkey-io/valkey/assets/51993843/888f2e47-6854-43e3-913e-35ef6aaea023)

Here is the link for what the version.script is https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_node/ld_25.html

**Comment by madolson** [comment]
I guess I'm still a bit confused, is there a test you can write that will demonstrate the behavior?

**Comment by PingXie** [comment]
The module in question is built against the Valkey source tree where `RedisModule_OnLoad` is aliased to `ValkeyModule_OnLoad`. The version script says that all symbols except for `RedisModule_OnLoad` should be hidden, which includes `ValkeyModule_OnLoad` hence this issue. Since we have decided to be both source and ABI compatiable with Redis modules, I think we can try fixing this issue by deleting https://github.com/valkey-io/valkey/blob/unstable/src/redismodule.h#L339

![image](https://github.com/valkey-io/valkey/assets/11568491/d92a2e33-3753-449a-855d-b35217c44156)

However, we will print out a "Legacy module ... loaded" NOTICE message with this change, even if the module is rebuilt against the Valkey tree. This is probably acceptable as the module is written against `redismodule.h`. So technically it is still a legacy Redis module.

**Comment by hwware** [comment]
> I guess I'm still a bit confused, is there a test you can write that will demonstrate the behavior?

Let me try to describe this issue by using an example

**Comment by hwware** [comment]
> I guess I'm still a bit confused, is there a test you can write that will demonstrate the behavior?

I already update the top comment to describe an example for this issue, Thanks

**Comment by PingXie** [comment]
@hwware the macro on L339 in redismodule.h is the cultprit.

**Comment by hwware** [comment]
> @hwware the macro on L339 in redismodule.h is the cultprit.

Yes, thus any client has the file like this, client need to be updated by itself.

**Comment by PingXie** [comment]
> Yes, thus any client has the file like this, client need to be updated by itself.

I wonder if we should take a code fix to remove this line altogether. If someone is bringing over a Redis module in the source code format, I think we should honor that intent by **not** changing their entry point. thoughts?

**Comment by madolson** [comment]
@hwware How critical is this to fix? Do we need to pull it in to the Valkey 8 project?

**Comment by hwware** [comment]
I have no idea how it is important. But any project includes vesion.script as the example in the top comment, some codes changes need. Let us have a vote for core member @valkey-io/core-team : do we need to pull it in to the Valkey 8 project?

**Comment by madolson** [comment]
I see nobody else responded. I'll add it to our next agenda just so we're tracking it and decide what to do about it.
